databuilder/query_engines/base_sql.py

```diff
@@ -220,6 +220,25 @@ def get_sql_subtract(self, node):
     def get_sql_multiply(self, node):
         return operators.mul(self.get_expr(node.lhs), self.get_expr(node.rhs))
 
+    @get_sql.register(Function.TrueDivide)
+    def get_sql_truedivide(self, node):
+        lhs = self.get_expr(node.lhs)
+        rhs = self.get_expr(node.rhs)
+        # To ensure TrueDiv behaviour cast one to float if both args are ints
+        if isinstance(lhs.type, sqlalchemy_types.Integer) and isinstance(
+            rhs.type, sqlalchemy_types.Integer
+        ):
+            lhs = sqlalchemy.cast(lhs, sqlalchemy.Float)
```

Comment:
There are a few things I've realised here since I first suggested this code. One is that the `isinstance()` check isn't actually enough because it doesn't work with custom types. So you have to use the `epxr_has_type` function from `sqlalchemy_query_utils`.

I've also realised that it's safer to check that values are _not_ floats, rather than check that they _are_ ints. If for some reason the type check doesn't work as expected, it's better to do an unnecessary CAST than to fail to CAST where we need to.

You can see both these at work here (which is where I realised this stuff):
https://github.com/opensafely-core/databuilder/blob/21c0c95d3dbf728e32d87ddfaf741648794df9a6/databuilder/query_engines/mssql.py#L33-L38

Reply:
Thanks, I can see how this is safer!